### PR TITLE
Optimize memory allocations in permissions cache

### PR DESCRIPTION
**Pull request by jalevin**
**What is this feature?**
This PR reduces the number of allocations made while caching permissions from the database.

**Why do we need this feature?**
Existing code allocates 4 times resulting in 4x memory usage in large datasets.

**Who is this feature for?**
Everyone

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by hairyhenderson** [comment]
I'm going to un-format this to make it easier to review 😅

File: pkg/services/accesscontrol/acimpl/service.go [hunk 1]
```diff
@@ -240,19 +244,16 @@ func (s *Service) getCachedUserPermissions(ctx context.Context, user identity.Re
...
-	permissions := make([]accesscontrol.Permission, 0, len(basicRolesPermissions)+len(teamsPermissions)+len(userPermissions))
-	permissions = append(permissions, basicRolesPermissions...)
-	permissions = append(permissions, teamsPermissions...)
 	permissions = append(permissions, userPermissions...)
+
 	return permissions, nil
 }
 
-func (s *Service) getCachedBasicRolesPermissions(ctx context.Context, user identity.Requester, options accesscontrol.Options) ([]accesscontrol.Permission, error) {
+func (s *Service) getCachedBasicRolesPermissions(ctx context.Context, user identity.Requester, options accesscontrol.Options, basicRolesPermissions []accesscontrol.Permission) ([]accesscontrol.Permission, error) {
```
Comment (hairyhenderson) [inline pkg/services/accesscontrol/acimpl/service.go:252]: FYI @jalevin I renamed this from `perms` so it causes no diffs inside the method body

**Review by gamab — APPROVED** [review]
LGTM, thanks! Same should be done in the enterprise code base :)

File: pkg/services/accesscontrol/acimpl/service.go (hunk 1 above)
Comment (gamab) [reply pkg/services/accesscontrol/acimpl/service.go:252]: @hairyhenderson I'd personally find it a bit more clear if we had permissions there :grin: 
```suggestion
func (s *Service) getCachedBasicRolesPermissions(ctx context.Context, user identity.Requester, options accesscontrol.Options, permissions []accesscontrol.Permission) ([]accesscontrol.Permission, error) {
```

File: pkg/services/accesscontrol/acimpl/service.go [hunk 2]
```diff
@@ -240,19 +244,16 @@ func (s *Service) getCachedUserPermissions(ctx context.Context, user identity.Re
 		return nil, err
 	}
 
-	permissions := make([]accesscontrol.Permission, 0, len(basicRolesPermissions)+len(teamsPermissions)+len(userPermissions))
-	permissions = append(permissions, basicRolesPermissions...)
-	permissions = append(permissions, teamsPermissions...)
 	permissions = append(permissions, userPermissions...)
+
```
Comment (hairyhenderson) [inline pkg/services/accesscontrol/acimpl/service.go:248]: It'd be interesting to record the # of permissions for posterity, how about we add it as a trace attribute:

```suggestion
span.SetAttributes(attribute.Int("num_permissions", len(permissions)))

```

File: pkg/services/accesscontrol/acimpl/service.go (hunk 1 above)
Comment (hairyhenderson) [reply pkg/services/accesscontrol/acimpl/service.go:252]: @gamab sure, but then line 261 would shadow it and it'd be confusing 😓

My goal was to reduce the diff to be as small as possible, but it's fair to rename too for clarity I guess...

File: pkg/services/accesscontrol/acimpl/service.go (hunk 2 above)
Comment (hairyhenderson) [reply pkg/services/accesscontrol/acimpl/service.go:249]: need to import `"go.opentelemetry.io/otel/attribute"` too 😅

File: pkg/services/accesscontrol/acimpl/service.go (hunk 1 above)
Comment (jalevin) [reply pkg/services/accesscontrol/acimpl/service.go]: I renamed to perms in function body (short lived var) and permissions as var passed in

File: pkg/services/authn/authnimpl/service.go
Comment (hairyhenderson) [inline pkg/services/authn/authnimpl/service.go:1]: can probably revert this, only changes are formatting

**Comment by hairyhenderson** [comment]
CI failing with:
```
       --- FAIL: TestDashboardAPIEndpoint/Given_provisioned_dashboard/When_calling_GET_on_/api/dashboards/uid/dash (0.01s)
        dashboard_service_mock.go:498: FAIL:	GetDashboard(string,mock.anythingOfTypeArgument)
            		at: [/drone/src/pkg/api/dashboard_test.go:648]
        dashboard_service_mock.go:498: FAIL: 0 out of 1 expectation(s) were met.
            	The code you are testing needs to make 1 more call(s).
            	at: [/drone/src/pkg/services/dashboards/dashboard_service_mock.go:498 /usr/local/go/src/testing/testing.go:1175 /usr/local/go/src/testing/testing.go:1353 /usr/local/go/src/testing/testing.go:1626 /usr/local/go/src/testing/testing.go:1634 /usr/local/go/src/runtime/panic.go:770 /usr/local/go/src/runtime/panic.go:261 /usr/local/go/src/runtime/signal_unix.go:881 /drone/src/pkg/api/dashboard.go:85 /drone/src/pkg/api/common_test.go:65 /drone/src/pkg/api/routing/routing.go:19 /drone/src/pkg/api/response/web_hack.go:40 /usr/local/go/src/net/http/server.go:2166 /drone/src/pkg/web/macaron.go:131 /usr/local/go/src/net/http/server.go:2166 /drone/src/pkg/services/contexthandler/contexthandler.go:141 /usr/local/go/src/net/http/server.go:2166 /drone/src/pkg/web/render.go:51 /usr/local/go/src/net/http/server.go:2166 /drone/src/pkg/web/context.go:52 /drone/src/pkg/web/router.go:155 /drone/src/pkg/web/router.go:221 /drone/src/pkg/web/macaron.go:163 /drone/src/pkg/api/common_test.go:175 /drone/src/pkg/api/dashboard_test.go:873 /drone/src/pkg/api/dashboard_test.go:860 /drone/src/pkg/api/dashboard_test.go:657 /drone/src/pkg/api/common_test.go:77]
        store_mock.go:662: FAIL:	GetProvisionedDataByDashboardID(string,mock.anythingOfTypeArgument)
            		at: [/drone/src/pkg/api/dashboard_test.go:641]
        store_mock.go:662: FAIL: 0 out of 1 expectation(s) were met.
            	The code you are testing needs to make 1 more call(s).
            	at: [/drone/src/pkg/services/dashboards/store_mock.go:662 /usr/local/go/src/testing/testing.go:1175 /usr/local/go/src/testing/testing.go:1353 /usr/local/go/src/testing/testing.go:1626 /usr/local/go/src/testing/testing.go:1634 /usr/local/go/src/runtime/panic.go:770 /usr/local/go/src/runtime/panic.go:261 /usr/local/go/src/runtime/signal_unix.go:881 /drone/src/pkg/api/dashboard.go:85 /drone/src/pkg/api/common_test.go:65 /drone/src/pkg/api/routing/routing.go:19 /drone/src/pkg/api/response/web_hack.go:40 /usr/local/go/src/net/http/server.go:2166 /drone/src/pkg/web/macaron.go:131 /usr/local/go/src/net/http/server.go:2166 /drone/src/pkg/services/contexthandler/contexthandler.go:141 /usr/local/go/src/net/http/server.go:2166 /drone/src/pkg/web/render.go:51 /usr/local/go/src/net/http/server.go:2166 /drone/src/pkg/web/context.go:52 /drone/src/pkg/web/router.go:155 /drone/src/pkg/web/router.go:221 /drone/src/pkg/web/macaron.go:163 /drone/src/pkg/api/common_test.go:175 /drone/src/pkg/api/dashboard_test.go:873 /drone/src/pkg/api/dashboard_test.go:860 /drone/src/pkg/api/dashboard_test.go:657 /drone/src/pkg/api/common_test.go:77]
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x3e34de1]

goroutine 3844 [running]:
testing.tRunner.func1.2({0x44bbc20, 0x82c0870})
	/usr/local/go/src/testing/testing.go:1631 +0x24a
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1634 +0x377
panic({0x44bbc20?, 0x82c0870?})
	/usr/local/go/src/runtime/panic.go:770 +0x132
github.com/grafana/grafana/pkg/api.(*HTTPServer).GetDashboard(0xc003020a08, 0xc00290c280)
	/drone/src/pkg/api/dashboard.go:85 +0xc1
github.com/grafana/grafana/pkg/api.loggedInUserScenarioWithRole.func1.1(0x44bd780?)
	/drone/src/pkg/api/common_test.go:65 +0xd5
github.com/grafana/grafana/pkg/api.loggedInUserScenarioWithRole.func1.Wrap.2(0xc00290c280)
	/drone/src/pkg/api/routing/routing.go:19 +0x1c
github.com/grafana/grafana/pkg/api/response.wrap_handler.func3({0x58a2088?, 0xc00095f770?}, 0x0?)
	/drone/src/pkg/api/response/web_hack.go:40 +0x30
net/http.HandlerFunc.ServeHTTP(...)
	/usr/local/go/src/net/http/server.go:2166
github.com/grafana/grafana/pkg/web.mwFromHandler.func1.1({0x58a2088?, 0xc00095f770?}, 0xc0014f1b00)
	/drone/src/pkg/web/macaron.go:131 +0x139
net/http.HandlerFunc.ServeHTTP(0xc003bd0000?, {0x58a2088?, 0xc00095f770?}, 0xc001d02f00?)
	/usr/local/go/src/net/http/server.go:2166 +0x29
github.com/grafana/grafana/pkg/services/contexthandler.(*ContextHandler).Middleware-fm.(*ContextHandler).Middleware.func1({0x58a2088, 0xc00095f770}, 0xc0014f1b00)
	/drone/src/pkg/services/contexthandler/contexthandler.go:141 +0xad5
net/http.HandlerFunc.ServeHTTP(0xc0011313e0?, {0x58a2088?, 0xc00095f770?}, 0xc002f0d580?)
	/usr/local/go/src/net/http/server.go:2166 +0x29
github.com/grafana/grafana/pkg/web.Renderer.func1.1({0x58a2088, 0xc00095f770}, 0xc0014f1b00)
	/drone/src/pkg/web/render.go:51 +0x26a
net/http.HandlerFunc.ServeHTTP(0x82d52f0?, {0x58a2088?, 0xc00095f770?}, 0xc0014d1f80?)
	/usr/local/go/src/net/http/server.go:2166 +0x29
github.com/grafana/grafana/pkg/web.(*Context).run(0xc00278de40)
	/drone/src/pkg/web/context.go:52 +0xdb
github.com/grafana/grafana/pkg/web.(*Router).Handle.func1({0x58a18a8, 0xc00278de00}, 0x4c06f88?, 0x2?)
	/drone/src/pkg/web/router.go:155 +0x14e
github.com/grafana/grafana/pkg/web.(*Router).ServeHTTP(0xc002dc8e00, {0x58a18a8, 0xc00278de00}, 0xc0014f1200)
	/drone/src/pkg/web/router.go:221 +0x195
github.com/grafana/grafana/pkg/web.(*Macaron).ServeHTTP(0xc0025c4750?, {0x58a18a8?, 0xc00278de00?}, 0xc0014f1200?)
	/drone/src/pkg/web/macaron.go:163 +0xdd
github.com/grafana/grafana/pkg/api.(*scenarioContext).exec(...)
	/drone/src/pkg/api/common_test.go:175
github.com/grafana/grafana/pkg/api.(*HTTPServer).callGetDashboard(0xc003020a08, 0xc0027fd040)
	/drone/src/pkg/api/dashboard_test.go:873 +0x133
github.com/grafana/grafana/pkg/api.getDashboardShouldReturn200WithConfig(_, _, {_, _}, {_, _}, {_, _}, {0x0, 0x0})
	/drone/src/pkg/api/dashboard_test.go:860 +0x95e
github.com/grafana/grafana/pkg/api.TestDashboardAPIEndpoint.func6.1(0xc0027fd040)
	/drone/src/pkg/api/dashboard_test.go:657 +0xf8
github.com/grafana/grafana/pkg/api.loggedInUserScenarioWithRole.func1(0xc003ce7d40?)
	/drone/src/pkg/api/common_test.go:77 +0x31a
testing.tRunner(0xc003ce7d40, 0xc0019f2420)
	/usr/local/go/src/testing/testing.go:1689 +0xfb
created by testing.(*T).Run in goroutine 3843
	/usr/local/go/src/testing/testing.go:1742 +0x390
```

**Review by hairyhenderson — APPROVED** [review]

**Comment by gamab** [comment]
> Same should be done in the enterprise code base :)

Hey :hand:, just wanted to make sure this had been accounted for :slightly_smiling_face:

**Comment by jalevin** [comment]
@gamab thanks! working on it now

**Comment by jalevin** [comment]
@gamab no problem! I've been out. I appreciate the contribution!

**Pull request merged**